HW2/P2/parallel_vector.pyx

```diff
@@ -54,39 +54,47 @@ cpdef move_data_serial(np.int32_t[:] counts,
                        np.int32_t[:] src,
                        np.int32_t[:] dest,
                        int repeat):
-   cdef:
-       int idx, r
+    cdef:
+        int idx, r
 
-   assert src.size == dest.size, "Sizes of src and dest arrays must match"
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+    assert src.size == dest.size, "Sizes of src and dest arrays must match"
+    with nogil:
+        for r in range(repeat):
+            for idx in range(src.shape[0]):
+                if counts[src[idx]] > 0:
+                    counts[dest[idx]] += 1
+                    counts[src[idx]] -= 1
 
 
 cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
+    cdef:
+        int idx, r, src_idx, dest_idx
+        omp_lock_t *locks = get_N_locks(counts.shape[0])
 
    ##########
    # Your code here
    # Use parallel.prange() and a lock for each element of counts to parallelize
    # data movement.  Be sure to avoid deadlock, and double-locking.
    ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+    with nogil:
+        for r in range(repeat):
+            for idx in prange(src.shape[0], num_threads=4):
+                src_idx, dest_idx = src[idx], dest[idx]
+                if counts[src_idx] > 0:
+                    # first increment dest[idx]
+                    acquire(&(locks[dest_idx]))
```

Comment:
I think you want to acquire the src lock before the if statement, because otherwise, you might be inside the execution under the assumption that the condition holds, but some other thread will have invalidated the condition in the meantime before you acquire the lock.
If you acquire the lock before the if statement and release it after, you will also have to deal with deadlocks by enforcing an order of which lock to acquire first